<#7> feat: interview, preanswer, comment controller & dto 설계

dadahee opened the pull request:
## 내용
- interview, preanswer, comment 도메인의 controller & dto 설계
- 해당 도메인 구현을 위한 일부 클래스의 디렉토리 구조 변경
- rest api를 위한 리팩토링

## 참고
- 다음 업무: #8

Commit:
feat: comment 컨트롤러/dto 작성
- CommentController dummy api 작성
- request dto validation을 위한 의존성 추가
- dto 추가 (comment는 update가 없으므로, CommentRequest 하나로 통일하여 작성)

Commit:
chore: comment controller/dto 수정
- parentresponse 제거

Commit:
feat: preAnswer 컨트롤러 / dto 작성
- PreAnswerController dummy api 작성
- request, response dto 추가
- comment: minor validation 코드 추가

Commit:
refactor: preAnswer, comment 내부 구조 수정, dto 변경
- comment, preanswer에서 사용되는 user response dto를 global dto로 변경
- save 시 201 status로 리턴하도록 변경
- service, serviceimpl로 분리
- repository, service 골격 추가

Commit:
chore: 주석 추가, response dto에서 createdAt 제거
- 임시 어노테이션 주석 추가
- CommentResponse dto에서 createdAt 제거

Commit:
chore: .gitkeep 제거, 주석 추가
- 도메인 폴더의 필요없는 `.gitkeep` 파일 제거
- 주석 추가: comment 목록 조회 기능은 더보기 구현 시에만 필요함

Commit:
feat: interview controller & dto 설계
- interview controller 더미 데이터 생성 및 api 설계
- interview dto 설계: 요약(Response), 상세(DetailResponse)로 구분하여 작성
- 연관 도메인 네이밍: Interview[도메인]Response

Commit:
feat: interview request dto 추가, tag request/response 구조 변경

Commit:
chore: interview update 시 request body에서 받아오도록 수정

Commit:
refactor: save/update 분리하여 request dto 구성하도록 변경

Commit:
feat: tag service/repository 클래스 추가, 디렉토리 정리

Commit:
Merge branch 'develop' into feature/07-back-IPCcontrollers

# Conflicts:
#	back-end/src/main/java/com/j2kb5th/chippo/interview/domain/Interview.java

Commit:
chore: 타이핑 에러 수정, serviceimpl 빈 어노테이션 추가

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/comment/controller/CommentController.java:
```diff
@@ -24,7 +26,7 @@
 public class CommentController {
 
-    private final CommentService commentService;
+    private final CommentServiceImpl commentService;
```
구현 클래스보다는 인터페이스로 주입해서 관심사 분리해주는 게 좋을 것 같아요!

dadahee replied:
엥 이건 제가 잘못 썼네요 감사합니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/comment/controller/CommentController.java:
```diff
@@ -76,10 +78,12 @@
         User commentUser = new User(1L, "dobby_wanna_be_free@google.com", "12345678", "n년차도비", Role.ROLE_USER, Provider.NAVER, true, false, null);
-        Interview interview = new Interview(interviewId, interviewUser, "VM과 Container 차이에 대해 설명하세요.", "VM은 어려운 것이고 Container도 어려운 것이다.", true, null, null, null);
-        Comment comment = new Comment(1L, commentUser, "면접 분위기는 전체적으로 어땠나요?", interview, null, null);
+        Interview interview = new Interview(interviewId, interviewUser, "VM과 Container 차이에 대해 설명하세요.", "VM은 어려운 것이고 Container도 어려운 것이다.", null, true, null, null, null);
```
파라미터가 많으니 생성자에 @Builder 를 추가하여 사용하는 편이 좋을 것 같아요.
그리고 각 도메인 클래스들 보니까 id 값은 넣어줄 필요가 없을 것 같습니다. auto_increment 라서 저장될 때마다 자동으로 +1 돼서 들어가는 것으로 알고 있어요!

dadahee replied:
클래스 선언부마다 위에 `@builder`를 선언해두었는데, 구체적인 생성자 안으로 옮길까요?
그리고 현재 id값은 데이터 처리 목적이 아닌, 더미 데이터 생성을 위해 전달해준 값이라 로직 구현 시에는 제외할 예정입니다!
아무래도 코드가 섞여있어서 알아보기 힘드실 것 같네요 ㅎㅎ...

nrudev replied:
번거로우시겠지만 생성자 안으로 옮기는 편이 좋을 것 같습니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/dto/request/UpdatePreAnswerRequest.java:
```diff
@@ -0,0 +1,20 @@
+    private Long id;
+
+    @NotNull
+    private Long userId;
```
사전답변을 업데이트하는데 사용자 정보가 필요할 것 같지는 않아 보여요!

nrudev replied:
update시에 id와 userId는 필요 없을 것 같습니다! 해당 변수 때문에 pk인 id가 수정될 수 있기 때문에 content만 남겨도 좋을 것 같아요.

dadahee replied:
Save, Update 공통 사항으로 여기 한 번에 남길게요 :)
update 시 request dto의 id 값과 path variable의 id 값을 비교하여 값이 일치 && 처리 권한(*작성자 == 로그인 유저 일치) 존재 하는 경우만 처리하는 로직(* 아닐 시 각각 400, 403 에러)을 생각했는데, 아무래도 빼는 게 좋을까요?
유저 아이디는 세션에서 얻어올 수 있으니 빼도 상관 없을 것 같고, id도 빼면 save/update request를 통일하여 사용할 수 있는 장점이 있겠네요. 다만 validation을 제가 위 방법으로만 해봐서 id를 빼게 된다면 체크를 어떻게 해주어야 할지 감이 안 잡히네요.
이 경우 사용자가 요청한 uri를 신뢰할 수 있음을 판단하려면 어떻게 해야할까요?

nrudev replied:
validation 목적이라면 남겨도 상관은 없을 듯 합니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/comment/controller/dto/request/CommentRequest.java:
```diff
@@ -16,6 +16,9 @@
 public class CommentRequest {
 
+    @NotNull
+    private Long userId;
```
userId 대신 User 객체를, parentId 도 Comment 객체를 넣어주는 게 맞지 않을까요?

dadahee replied:
사실 commentRequest는 회의 때도 다시 말씀드리겠지만, child가 parent를 품는 구조가 아니라 parent가 children 배열을 품는 구조로 프론트에 쏴줘야 하지 않나? 생각이 듭니다. parent 정보만으로는 덧글의 위계 파악이 어려울 것 같아서요. 1계층이니 말씀해주신 것처럼 parent를 객체로 쏴주면 대공사 하지 않는 선에서 고칠 수 있을 것 같긴 한데... 준승님께도 여쭤봐야 할 것 같아요.

nrudev replied:
이거는 오늘 회의 때 말씀 나누는 걸로 하시죠ㅎㅎ

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/dto/response/PreAnswersResponse.java:
```diff
@@ -0,0 +1,12 @@
+
+@Getter
+@RequiredArgsConstructor // 임시용
+public class PreAnswersResponse {
```
PreAnswerResponse 가 있어서 조금 헷갈릴 것 같은데 차이가 좀 분명한 명칭으로 바꾸는 것이 좋지 않을까요?

dadahee replied:
추천하시거나 그동안 사용하셨던 명칭이 있을까요? 당장은 `PreAnswerListResponse` 정도만 생각나네요..

nrudev replied:
네 저도 그 정도면 적당한 듯 합니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/dto/response/PreAnswerResponse.java:
```diff
@@ -1,12 +1,16 @@
     private final String content;
-    private final PreAnswerUserResponse user;
+    private final UserResponse user;
+    private final LocalDateTime updatedAt; // 논의 필요
```
PreAnswer에는 다희님이 작성하신 그대로 updatedAt으로 해도 괜찮을 듯 합니다! 👍

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/dto/request/SavePreAnswerRequest.java:
```diff
@@ -6,7 +6,10 @@
+public class SavePreAnswerRequest {
+
+    @NotNull
+    private Long userId;
```
db에서는 유저id 로만 보이긴 하는데, 도메인 클래스에 User 객체로 되어 있어서 User 를 넣어주어야 할 것 같습니다.

dadahee replied:
id 옵션을 User 객체로 래핑하여 보내자는 말씀이시죠?

```
{
    ...
    "user": {
        "id": number,
     }
}
```
이런 식으로요.

아니라면, 이 외에 request에서 User 객체로 래핑하여 받을 데이터가 무엇이 있을까요?
** PreAnswer Save/Update를 위해 꼭 필요한 데이터(id)가 아니라면(ex. email) 굳이 포함하지 않는 게 좋을 것 같습니다.

nrudev replied:
이거 회의때 화면공유하면서 의논하는 편이 좋을 것 같습니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/request/SaveInterviewRequest.java:
```diff
@@ -0,0 +1,35 @@
+    private String extraInfo;
+
+    @NotNull
+    private boolean visible;
```
visible 은 기본값으로 true 값을 주기 때문에 삭제해도 좋을 것 같습니다.

dadahee replied:
@NotNull을 없애도 된다는 말씀일까요? 아님 필드를 없애자는 말씀일까요?
후자라면 등록은 전체 공개로 하고, 수정 시 비공개 여부를 선택할 수 있는 구조가 되는 걸까요?

nrudev replied:
제가 visible 컬럼은 삭제 여부를 판단하는 컬럼으로 이해하고 있습니다.
그래서 저장할 때 기본값을 true로 주기 위해서 Interview 도메인 클래스에 @ColumnDefault와 함께 @DynamicInsert 어노테이션을 붙였던 것으로 알고 있어요.

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/request/SaveInterviewRequest.java:
```diff
@@ -0,0 +1,35 @@
+public class SaveInterviewRequest {
+
+    @NotNull
+    private Long userId;
```
이 또한 userId가 아니라 User 객체를 넣어주어야 하지 않을까 싶습니다!

dadahee replied:
위와 동일합니다!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/request/UpdateInterviewRequest.java:
```diff
@@ -0,0 +1,38 @@
+    private Long id;
+
+    @NotNull
+    private Long userId;
```
id나 userId 는 수정시에 필요한 데이터는 아닌 듯 합니다. 이게 있으면 자칫 잘못하면 pk인 interview_id 값이나 해당 게시물 작성자가 변경될 수 있을 것 같아요.

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/request/UpdateInterviewRequest.java:
```diff
@@ -0,0 +1,38 @@
+    private String extraInfo;
+
+    @NotNull
+    private boolean visible;
```
visible 은 삭제시에만 업데이트되면 되기 때문에 따로 DeleteInterviewRequest 로 빼는 편이 좋을 것 같습니다.

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/request/UpdateInterviewTagDetailRequest.java:
```diff
@@ -0,0 +1,23 @@
+public class UpdateInterviewTagDetailRequest {
+
+    @NotNull
+    private Long id;
```
id 는 수정될 필요가 없는 데이터기 때문에 빼도 좋을 것 같습니다.

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/response/InterviewDetailResponse.java:
```diff
@@ -18,10 +18,10 @@
     private final String answer;
-    private final String extra_info;
+    private final String extraInfo;
     private final boolean visible;
```
조회시에 필요한 DTO 인데, 애초에 visible이 false 면 조회되지 않도록 할 생각이기 때문에 visible을 삭제해도 되지 않을까 싶습니다.

dadahee replied:
아하, visible이 열람 제한 여부(*작성자 포함)였군요? 저는 여태 사용자가 임시 등록처럼 <나만 보기> 식으로 등록할 수 있는 건줄 알았네요.  아예 네이밍을 바꿔야 하지 않을까 싶습니다 ^^;;

nrudev replied:
해당 컬럼이 삭제여부를 결정하는 컬럼으로 기억하는데, 아니었나요..?

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/dto/response/InterviewsResponse.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@RequiredArgsConstructor // 구현 후 삭제
+public class InterviewsResponse {
```
InterviewResponse 가 있어서 조금 헷갈릴 것 같은데 InterviewListResponse 같이 차이가 좀 분명한 명칭으로 바꾸는 것이 좋지 않을까요?

dadahee replied:
오~ 찌찌뽕입니다. `~ListResponse`로 바꿀게요!!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/InterviewController.java:
```diff
@@ -0,0 +1,170 @@
+
+    @PutMapping
+    public ResponseEntity<InterviewDetailResponse> updateInterview(
+            @PathVariable(name = "interviewId") Long interviewId
```
PathVariable이 있는데, 맵핑된 주소에 interviewId 가 적혀있질 않아요!

nrudev commented on back-end/src/main/java/com/j2kb5th/chippo/interview/controller/InterviewController.java:
```diff
@@ -0,0 +1,170 @@
+
+    @DeleteMapping
+    public ResponseEntity<Void> deleteInterview(
+            @PathVariable(name = "interviewId") Long interviewId
```
PathVariable이 있는데, 맵핑된 주소에 interviewId 가 적혀있질 않아요!

nrudev reviewed (commented):
고생 많으셨습니다! 제가 틀린 사항도 있을 수 있으니 리뷰한 것들 확인하여 답글 부탁드립니다~

Commit:
refactor: interview, preAnswer, comment 코드리뷰 및 회의 내용 반영
- 엔티티에서 불필요한 필드 삭제
- PathVariable 누락 수정
- PreAnswer 조회 및 작성 방법 변경
- 빌더 패턴 반영
- Like -> Thumb 변경
- Interview 조회 시 데이터 추가 및 변경
    - thumb: 로그인한 사용자의 thumb 여부, 게시글 thumb count
    - PreAnswer: 컨트롤러에서 목록 조회 삭제, 단건 조회(임시용) 추가

Commit:
chore: interview, preAnswer, comment 빌더 패턴 적용

nrudev commented:
고생 많으셨습니다!
